# Chan Kai Yang
# 7372711

## Importing Necessary Libraries

In [1]:
# import numpy, pandas libraries
import pandas as pd
import numpy as np

import random
import math

#For pretty formatting print
from pprint import pprint

## Pre-prossessing
### Reading data file

In [2]:
#Add in headers for the data frame
#Ensure sep =', ' as there is a space before each feature
df = pd.read_csv("adult.data", sep=", ", engine='python',
                  names=["age", "workclass", "fnlwgt", "education", "education-num", "marital-status", "occupation", "race", "relationship", "sex", "capital-gain", "capital-loss", "hours-per-week", "native-country", "income"])

print(df)

df.head(30)

       age         workclass  fnlwgt   education  education-num  \
0       39         State-gov   77516   Bachelors             13   
1       50  Self-emp-not-inc   83311   Bachelors             13   
2       38           Private  215646     HS-grad              9   
3       53           Private  234721        11th              7   
4       28           Private  338409   Bachelors             13   
...    ...               ...     ...         ...            ...   
32556   27           Private  257302  Assoc-acdm             12   
32557   40           Private  154374     HS-grad              9   
32558   58           Private  151910     HS-grad              9   
32559   22           Private  201490     HS-grad              9   
32560   52      Self-emp-inc  287927     HS-grad              9   

           marital-status         occupation           race relationship  \
0           Never-married       Adm-clerical  Not-in-family        White   
1      Married-civ-spouse    Exec-manageria

age         workclass  fnlwgt     education  education-num  \
0    39         State-gov   77516     Bachelors             13   
1    50  Self-emp-not-inc   83311     Bachelors             13   
2    38           Private  215646       HS-grad              9   
3    53           Private  234721          11th              7   
4    28           Private  338409     Bachelors             13   
5    37           Private  284582       Masters             14   
6    49           Private  160187           9th              5   
7    52  Self-emp-not-inc  209642       HS-grad              9   
8    31           Private   45781       Masters             14   
9    42           Private  159449     Bachelors             13   
10   37           Private  280464  Some-college             10   
11   30         State-gov  141297     Bachelors             13   
12   23           Private  122272     Bachelors             13   
13   32           Private  205019    Assoc-acdm             12   
14   40           Private  121772     Assoc-voc             11   
15   34           Private  245487       7th-8th              4   
16   25  Self-emp-not-inc  176756       HS-grad              9   
17   32           Private  186824       HS-grad              9   
18   38           Private   28887          11th              7   
19   43  Self-emp-not-inc  292175       Masters             14   
20   40           Private  193524     Doctorate             16   
21   54           Private  302146       HS-grad              9   
22   35       Federal-gov   76845           9th              5   
23   43           Private  117037          11th              7   
24   59           Private  109015       HS-grad              9   
25   56         Local-gov  216851     Bachelors             13   
26   19           Private  168294       HS-grad              9   
27   54                 ?  180211  Some-college             10   
28   39           Private  367260       HS-grad              9   
29   49           Private  193366       HS-grad              9   

           marital-status         occupation           race  \
0           Never-married       Adm-clerical  Not-in-family   
1      Married-civ-spouse    Exec-managerial        Husband   
2                Divorced  Handlers-cleaners  Not-in-family   
3      Married-civ-spouse  Handlers-cleaners        Husband   
4      Married-civ-spouse     Prof-specialty           Wife   
5      Married-civ-spouse    Exec-managerial           Wife   
6   Married-spouse-absent      Other-service  Not-in-family   
7      Married-civ-spouse    Exec-managerial        Husband   
8           Never-married     Prof-specialty  Not-in-family   
9      Married-civ-spouse    Exec-managerial        Husband   
10     Married-civ-spouse    Exec-managerial        Husband   
11     Married-civ-spouse     Prof-specialty        Husband   
12          Never-married       Adm-clerical      Own-child   
13          Never-married              Sales  Not-in-family   
14     Married-civ-spouse       Craft-repair        Husband   
15     Married-civ-spouse   Transport-moving        Husband   
16          Never-married    Farming-fishing      Own-child   
17          Never-married  Machine-op-inspct      Unmarried   
18     Married-civ-spouse              Sales        Husband   
19               Divorced    Exec-managerial      Unmarried   
20     Married-civ-spouse     Prof-specialty        Husband   
21              Separated      Other-service      Unmarried   
22     Married-civ-spouse    Farming-fishing        Husband   
23     Married-civ-spouse   Transport-moving        Husband   
24               Divorced       Tech-support      Unmarried   
25     Married-civ-spouse       Tech-support        Husband   
26          Never-married       Craft-repair      Own-child   
27     Married-civ-spouse                  ?        Husband   
28               Divorced    Exec-managerial  Not-in-family   
29     Married-civ-spouse       Craft-repair        Husband   

   

In [3]:
#Check for null values or missing values
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32561 entries, 0 to 32560
Data columns (total 15 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   age             32561 non-null  int64 
 1   workclass       32561 non-null  object
 2   fnlwgt          32561 non-null  int64 
 3   education       32561 non-null  object
 4   education-num   32561 non-null  int64 
 5   marital-status  32561 non-null  object
 6   occupation      32561 non-null  object
 7   race            32561 non-null  object
 8   relationship    32561 non-null  object
 9   sex             32561 non-null  object
 10  capital-gain    32561 non-null  int64 
 11  capital-loss    32561 non-null  int64 
 12  hours-per-week  32561 non-null  int64 
 13  native-country  32561 non-null  object
 14  income          32561 non-null  object
dtypes: int64(6), object(9)
memory usage: 3.7+ MB


In [4]:
#Notice workclass and occupation has a '?' value on the same row (row 27)
#Notice native-country has a '?' value on row 14
#Drop all rows with ' ?' values as they are not in the attribute information
#irrelevant as we cannot get any information about workclass or occupation from ?

#Loop through all columns in dataframe and remove the rows with '?'
for column in df:
    df.drop(df.index[df[column] == '?'], inplace=True)

#Check that all rows with '?' have been removed
df.head(30)

#Row 27 is removed
#Row 14 is removed

age         workclass  fnlwgt     education  education-num  \
0    39         State-gov   77516     Bachelors             13   
1    50  Self-emp-not-inc   83311     Bachelors             13   
2    38           Private  215646       HS-grad              9   
3    53           Private  234721          11th              7   
4    28           Private  338409     Bachelors             13   
5    37           Private  284582       Masters             14   
6    49           Private  160187           9th              5   
7    52  Self-emp-not-inc  209642       HS-grad              9   
8    31           Private   45781       Masters             14   
9    42           Private  159449     Bachelors             13   
10   37           Private  280464  Some-college             10   
11   30         State-gov  141297     Bachelors             13   
12   23           Private  122272     Bachelors             13   
13   32           Private  205019    Assoc-acdm             12   
15   34           Private  245487       7th-8th              4   
16   25  Self-emp-not-inc  176756       HS-grad              9   
17   32           Private  186824       HS-grad              9   
18   38           Private   28887          11th              7   
19   43  Self-emp-not-inc  292175       Masters             14   
20   40           Private  193524     Doctorate             16   
21   54           Private  302146       HS-grad              9   
22   35       Federal-gov   76845           9th              5   
23   43           Private  117037          11th              7   
24   59           Private  109015       HS-grad              9   
25   56         Local-gov  216851     Bachelors             13   
26   19           Private  168294       HS-grad              9   
28   39           Private  367260       HS-grad              9   
29   49           Private  193366       HS-grad              9   
30   23         Local-gov  190709    Assoc-acdm             12   
31   20           Private  266015  Some-college             10   

           marital-status         occupation           race  \
0           Never-married       Adm-clerical  Not-in-family   
1      Married-civ-spouse    Exec-managerial        Husband   
2                Divorced  Handlers-cleaners  Not-in-family   
3      Married-civ-spouse  Handlers-cleaners        Husband   
4      Married-civ-spouse     Prof-specialty           Wife   
5      Married-civ-spouse    Exec-managerial           Wife   
6   Married-spouse-absent      Other-service  Not-in-family   
7      Married-civ-spouse    Exec-managerial        Husband   
8           Never-married     Prof-specialty  Not-in-family   
9      Married-civ-spouse    Exec-managerial        Husband   
10     Married-civ-spouse    Exec-managerial        Husband   
11     Married-civ-spouse     Prof-specialty        Husband   
12          Never-married       Adm-clerical      Own-child   
13          Never-married              Sales  Not-in-family   
15     Married-civ-spouse   Transport-moving        Husband   
16          Never-married    Farming-fishing      Own-child   
17          Never-married  Machine-op-inspct      Unmarried   
18     Married-civ-spouse              Sales        Husband   
19               Divorced    Exec-managerial      Unmarried   
20     Married-civ-spouse     Prof-specialty        Husband   
21              Separated      Other-service      Unmarried   
22     Married-civ-spouse    Farming-fishing        Husband   
23     Married-civ-spouse   Transport-moving        Husband   
24               Divorced       Tech-support      Unmarried   
25     Married-civ-spouse       Tech-support        Husband   
26          Never-married       Craft-repair      Own-child   
28               Divorced    Exec-managerial  Not-in-family   
29     Married-civ-spouse       Craft-repair        Husband   
30          Never-married    Protective-serv  Not-in-family   
31          Never-married              Sales      Own-child   

   

In [5]:
#Check that all rows with '?' have been removed
df.info()

#Around 2k rows removed

<class 'pandas.core.frame.DataFrame'>
Int64Index: 30162 entries, 0 to 32560
Data columns (total 15 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   age             30162 non-null  int64 
 1   workclass       30162 non-null  object
 2   fnlwgt          30162 non-null  int64 
 3   education       30162 non-null  object
 4   education-num   30162 non-null  int64 
 5   marital-status  30162 non-null  object
 6   occupation      30162 non-null  object
 7   race            30162 non-null  object
 8   relationship    30162 non-null  object
 9   sex             30162 non-null  object
 10  capital-gain    30162 non-null  int64 
 11  capital-loss    30162 non-null  int64 
 12  hours-per-week  30162 non-null  int64 
 13  native-country  30162 non-null  object
 14  income          30162 non-null  object
dtypes: int64(6), object(9)
memory usage: 4.7+ MB


### Testing data using a copy of df

In [6]:
# Make a copy of the data for processing
df_copy = df.copy()
df_copy.head()

age         workclass  fnlwgt  education  education-num  \
0   39         State-gov   77516  Bachelors             13   
1   50  Self-emp-not-inc   83311  Bachelors             13   
2   38           Private  215646    HS-grad              9   
3   53           Private  234721       11th              7   
4   28           Private  338409  Bachelors             13   

       marital-status         occupation           race relationship     sex  \
0       Never-married       Adm-clerical  Not-in-family        White    Male   
1  Married-civ-spouse    Exec-managerial        Husband        White    Male   
2            Divorced  Handlers-cleaners  Not-in-family        White    Male   
3  Married-civ-spouse  Handlers-cleaners        Husband        Black    Male   
4  Married-civ-spouse     Prof-specialty           Wife        Black  Female   

   capital-gain  capital-loss  hours-per-week native-country income  
0          2174             0              40  United-States  <=50K  
1             0             0              13  United-States  <=50K  
2             0             0              40  United-States  <=50K  
3             0             0              40  United-States  <=50K  
4             0             0              40           Cuba  <=50K

### Implementing splitting function to split the dataset into a train and test set

In [7]:
# Split a dataset into a train and test set
def train_test_split(dataset, test_size):
    if isinstance(test_size,float):
        #calculate the number of rows for test set based on test size
        test_size = round(test_size * len(dataset))
    
    #All rows in df to list
    indices = dataset.index.tolist()
    
    # A random sample of rows added to the test set based on test_size.
    test_indices = random.sample(population=indices, k=test_size)
    
    #update indices list
    indices = [ele for ele in indices if ele not in test_indices]
    
    #access rows with test_indices
    test = dataset.loc[test_indices]
    
    #Access rows without test_indices
    train = dataset.drop(test_indices)
 
    return train, test

### Test out split function to see if proportions and randomization is correct

In [8]:
#rows of test_df should be different from those in train_df
#40% as test data, 60% as training data
random.seed(42)
train_df, test_df =  train_test_split(df_copy, test_size=0.4)

In [9]:
train_df.head()

#row 0, 1, 4, 7, 8

age         workclass  fnlwgt  education  education-num  \
0   39         State-gov   77516  Bachelors             13   
1   50  Self-emp-not-inc   83311  Bachelors             13   
4   28           Private  338409  Bachelors             13   
7   52  Self-emp-not-inc  209642    HS-grad              9   
8   31           Private   45781    Masters             14   

       marital-status       occupation           race relationship     sex  \
0       Never-married     Adm-clerical  Not-in-family        White    Male   
1  Married-civ-spouse  Exec-managerial        Husband        White    Male   
4  Married-civ-spouse   Prof-specialty           Wife        Black  Female   
7  Married-civ-spouse  Exec-managerial        Husband        White    Male   
8       Never-married   Prof-specialty  Not-in-family        White  Female   

   capital-gain  capital-loss  hours-per-week native-country income  
0          2174             0              40  United-States  <=50K  
1             0             0              13  United-States  <=50K  
4             0             0              40           Cuba  <=50K  
7             0             0              45  United-States   >50K  
8         14084             0              50  United-States   >50K

In [10]:
test_df.head()

#row 22609, 3977, 891, 26219, 9753

age         workclass  fnlwgt   education  education-num  \
22609   48           Private  143281     HS-grad              9   
3977    24           Private  228649   Bachelors             13   
891     35  Self-emp-not-inc  372525   Bachelors             13   
26219   42         State-gov  219682     Masters             14   
9753    44         Local-gov  107845  Assoc-acdm             12   

      marital-status         occupation           race relationship     sex  \
22609        Widowed  Machine-op-inspct  Not-in-family        White  Female   
3977   Never-married       Adm-clerical      Own-child        White    Male   
891    Never-married    Exec-managerial  Not-in-family        White    Male   
26219       Divorced    Exec-managerial  Not-in-family        White    Male   
9753        Divorced    Protective-serv  Not-in-family        White  Female   

       capital-gain  capital-loss  hours-per-week native-country income  
22609             0             0              48  United-States  <=50K  
3977              0             0              38  United-States  <=50K  
891               0             0              40  United-States  <=50K  
26219             0             0              45  United-States  <=50K  
9753              0             0              56  United-States   >50K

### Check that proportions are correct

In [11]:
#60% of data in training set
len(train_df)

18097

In [12]:
#40% of data in test set
len(test_df)

12065

In [13]:
#Check total length of set
len(df_copy)

30162

# Implementing the decision tree classifier


## 1. Function for calculating information gain (entropy)

In [14]:
# calculate entropy/info gain for the dataset

#Last column of dataframe is usually the label column
label_col  = train_df.iloc[:, -1]

#Find unique values and frequency in label column 
_, n_counts = np.unique(label_col, return_counts=True)

#Probability of unique values (frequency of value/sum of all frequencies)
prob = n_counts / n_counts.sum()

#Entropy = sum of (probability * -log2(probabilities))
entropy_of_dataset = sum(prob * -np.log2(prob))

print("Entropy of dataset: ", entropy_of_dataset)

Entropy of dataset:  0.803546825112663


#### Measuring impurity of dataset. An entropy of ~0.8 means that our dataset is relatively impure, and that our data is widely spread out among our classes.

In [15]:
#Function for calculating entropy
def calculate_entropy(data):
    
    #Last column of dataframe is usually the label column
    label_col  = data[:, -1]
    
    #Find unique values and frequency in label column 
    _, n_counts = np.unique(label_col, return_counts=True)

    #Probability of unique values (frequency of value/sum of all frequencies)
    prob = n_counts / n_counts.sum()
    
    #Entropy = sum of (probability * -log2(probabilities))
    entropy = sum(prob * -np.log2(prob))
     
    return entropy

### Function for calculating entropy of splits

In [16]:
def calculate_child_entropy(left, right):
    
    # calculate total number of rows
    n = len(left) + len(right)
    
    # calculate the occurences of each feature
    p_left = len(left) / n
    p_right = len(right) / n

    # calculate weighted entropy based on the probabilities of each feature
    child_entropy =  (p_left * calculate_entropy(left) + p_right * calculate_entropy(right))
    
    return child_entropy

## 2. Function for calculating gini impurity

### Function for calculating gini impurity for label

In [17]:
#calculate gini for target
def calculate_gini(data):
    
    #All unique class labels
    class_labels = np.unique(data)
    
    gini = 0
    
    for classes in class_labels:
        
        #probability of a record belonging to each class 
        p_classes = len(data[data == classes]) / len(data)
        
        #Add to gini impurity
        gini += p_classes**2
        
    return 1 - gini

### Function for calculating gini impurity for feature

In [18]:
def calculate_gini_gain(parent, left, right):
    
    #Weight of left node
    left_weight = len(left) / len(parent)
    
    #Weight of right node
    right_weight = len(right) / len(parent)
    
    #Gini gain = gini impurity of the parent node - the weighted sum of the gini impurities of the two child nodes
    giniGain = calculate_gini(parent) - (left_weight *calculate_gini(left) + right_weight *calculate_gini(right))
    
    return giniGain

#### The function calculates and returns the gini impurity gain, which is the improvement/reduction in homogeneity that results from splitting the "parent" data into two child nodes "left" and "right". 

## 3. Function for getting all possible splits

In [19]:
def get_possible_splits(data):
    possible_splits = {}
    
    #Get no. of rows n columns using shape property
    n_rows , n_cols = data.shape
    
    #Get all possible splits for each feature except targe
    for col_index in range(n_cols-1):
        possible_splits[col_index] = {}
        values = data[:, col_index]
        unique_values = np.unique(values)
    
        possible_splits[col_index] = unique_values
                
    return possible_splits

#### The "possible_splits" dictionary returned by the function contains all the possible split points for each feature column in the dataframe, which can be used to find the best split that maximizes the gini impurity/info/gain ratio gain.

### Function for binary splits

In [20]:
# split the dataframe into two child nodes produced by split
def split_data(data, split_col, split_val):
    
    split_col_val = data[:, split_col]
    
    left = data[split_col_val == split_val]
    right = data[split_col_val !=  split_val]
    
    return left, right

## Function for calculating best split based on information gain

In [21]:
#Function for calculating best split based on information gain
def determine_best_split_infogain(data, possible_splits):
    
    # calculate the entropy of the data
    parent_entropy = calculate_entropy(data)
    best_entropy = parent_entropy

    # loop through all values for each column
    for col_index in possible_splits:
        
        for value in possible_splits[col_index]:
            
            #Split data into 2 nodes
            left, right = split_data(data, split_col=col_index, split_val=value)
            
            #Calculate entropy of split
            child_entropy = calculate_child_entropy(left,right)
            
            # compare each entropy value and find lowest
            if child_entropy < best_entropy:
                best_split_col = col_index
                best_split_val = value
                best_entropy = child_entropy
                info_gain = parent_entropy - best_entropy
    
    return best_split_col, best_split_val

## Function for calculating best split based on gini index

In [22]:
def determine_best_split_gini_index(data, possible_splits):
    
    label_col  = data[:, -1]
  
    # initialise max gini for comparision
    max_gini_gain = -float("inf")

    
    #dictionary to store possible splits and corresponding gini index
    gini = {}
    
    # loop through all values for each column
    for col_index in possible_splits:
        for value in possible_splits[col_index]:
            
            left, right = split_data(data, split_col=col_index, split_val=value)
            if len(left)>0 and len(right)>0:            
                y, left_y, right_y = data[:, -1], left[:, -1], right[:, -1]
                #calculate gini for each feature subset
                curr_gini_gain = calculate_gini_gain(y, left_y, right_y)

                ##feature with highest gini gain is selected as the splitting feature
                # compare each subset's value
                if curr_gini_gain > max_gini_gain:
                    max_gini_gain = curr_gini_gain
                    best_split_col = col_index
                    best_split_val = value

    return best_split_col, best_split_val

## Function for calculating best split based on gain ratio

In [23]:
def determine_best_split_gain_ratio(data, possible_splits):
    #Calculate the entropy of the data
    parent_entropy = calculate_entropy(data)

    # Initialise best entropy and gain ratio
    best_entropy = parent_entropy
    best_gain_ratio = 0

    #Loop over each column and values in possible splits
    for col_index in possible_splits:
        
        for value in possible_splits[col_index]:
            #Split data into 2 nodes
            left, right = split_data(data, split_col=col_index, split_val=value)

            #Get child entropy for split
            child_entropy = calculate_child_entropy(left, right)

            #Get split info for the current split
            split_info = calculate_Split_info(possible_splits[col_index])

            #Compute info gain for split
            info_gain = parent_entropy - child_entropy

            # Compute the gain ratio for the split
            gain_ratio = info_gain / (split_info + 1e-10)

            #compare child entropy
            #if child_entropy < best_entropy AND
            #compare gain ratio
            #feature with highest gain ratio will be used to split
            if child_entropy < best_entropy and gain_ratio > best_gain_ratio:
                best_entropy = child_entropy
                best_gain_ratio = gain_ratio
                best_split_col = col_index
                best_split_val = value

    return best_split_col, best_split_val

In [24]:
# Function to calculate split information for one feature
def calculate_Split_info(data_col):
    
    # Initialize the split info
    split_info = 0
    
    # Get the unique values and the count of each value in feature
    unique_vals, n_counts = np.unique(data_col, return_counts=True)
    
    # Probability of each unique value
    prob = n_counts / n_counts.sum()
    
    for idx, values in enumerate(unique_vals):

        # Calculate the number of occurrences of the unique value
        occurences = n_counts[idx]
        
        prob = occurences/n_counts.sum()
        temp = (prob* - np.log2(prob))
        split_info = split_info + temp

    return(split_info)

## Function for checking if we have reached a leaf node

In [25]:
def check_leaf(data):
    #store last column as label_col
    label_col = data[:,-1]
    
    #extract unique values from label_col
    classes = np.unique(label_col)

    # If there is only one unique value, the data is a leaf node
    if len(classes) == 1:
        return True
    else:
        return False

## Function for classification

In [26]:
#function gets the class for when a leaf node is reached
def classify_data(data):
    
    #since labels are always in the last column
    label_col = data[:,-1]
    
    #find each unique classes in the label column and how many times it appears
    classes, n_classes = np.unique(label_col, return_counts=True)

    #find index of the most frequent class
    index = n_classes.argmax()
    
    #extract out the name
    classification = classes[index]
    
    return classification

In [27]:
#Test function
classify_data(train_df.values)

'<=50K'

## Function for predicting values using the tree

In [28]:
#Get label prediction for 1 row
def classify_one(test, tree):
    
    # save tree dictionary into list
    #question = keys
    question = list(tree.keys())[0]
    
    # extract out feature label, operator'=' and values of the feature
    # feature_name = key, operator = "=", value = value of key
    feature_name, operator, value = question.split(" ")

    # asking the question
    #if the sample value matches the question's value, it moves on to the left subtree, otherwise it moves to the right subtree. 
    if str(test[feature_name]) == value:
        answer = tree[question][0]
    else:
        answer = tree[question][1]

   
    #if it is a dictionary, it means that the answer is a subtree.
    #recursively call method until answer is no longer a dictionary, and hence found
    #else, it is the answer
    #then return answer as final prediction
    
    #If not dictionary return ans as prediction
    if not isinstance(answer, dict):
        return answer
    
    #Else recursively call function until ans is found
    else:
        residual_tree = answer
        return classify_one(test, residual_tree)

In [29]:
#Get a list of predictions for all rows in test_df
def predict(test,tree):
    #initilaise array for storing 
    output = []
    index = test.index.tolist()
    
    #Loop through each row of test_df and apply classify_one function to predict label for row
    for i in index:
        output.append(classify_one(test.loc[i],tree))
        
    #Returns a list of predictions for each row in test_df.
    return output

## Function for calculating prediction accuracy

In [30]:
#By comparing all predictions to its actual values, calculating the average mean of total predictions
def calculate_accuracy(df, tree):

    #Add a new column to df with the result of the prediction of each row using the "classify_one" function
    df["classification"] = df.apply(classify_one, axis=1, args=(tree,))
    
    #If classification for row is correct, add it to df["classification_correct"]
    df["classification_correct"] = df["classification"] == df["income"]
    
    # Calculate the accuracy using mean of total predictions
    accuracy = df["classification_correct"].mean()
    
    return accuracy

## Splitting dataset into training (60%) and testing sets (40%).

In [31]:
#Split train (60%) and test set (40%)
train_df, test_df = train_test_split(df, test_size=0.4)

## 1. decision tree classifier based on infogain

In [32]:
#parameters
#df: the dataframe containing the data to be trained and classified
#max_depth: the maximum depth of the decision tree/the maximum number of splits
#min_samples: the minimum number of samples required for a split to occur
#counter: a counter variable to keep track of the depth of the tree
def decision_tree_classifier_infogain(df, max_depth, min_samples, counter=0):
    
    #initialize variables at the start
    if counter == 0:
        global COLUMN_HEADERS
        COLUMN_HEADERS = df.columns
        data = df.values
    else:
        data = df           
    
    # early stopping conditions
    # if the data is a leaf node/min samples > number of samples in data/max depth reached = get leaf node classification
    if (check_leaf(data)) or (len(data) < min_samples) or (counter == max_depth):
        classification = classify_data(data)
        
        return classification
    
    #Else, get all possible splits and choose the best split with best info gain
    #Then it splits the data into two branches, left and right, based on the best split and the question generated from it.
    else:    
        # increment the counter for each iteration
        counter += 1
        # get all possible splits
        possible_splits = get_possible_splits(data)
        
        #returns the feature/col with best info gain
        split_column, split_value = determine_best_split_infogain(data, possible_splits)
        
        #split data into left and right branches using best feature as decided by info gain
        left, right = split_data(data, split_column, split_value)
        
        # choose question (key)
        # get the name of the feature that was used for the best split
        feature_name = COLUMN_HEADERS[split_column]
            
        #format question (key) (e.g. "feature_name = split_value")
        question = "{} = {}".format(feature_name, split_value)
        
        #initialise dictionary for storing tree
        sub_tree = {question: []}
    
        #recursive call with left and right branches
        yes_answer = decision_tree_classifier_infogain(left, max_depth , min_samples, counter)
        no_answer = decision_tree_classifier_infogain(right, max_depth, min_samples, counter)
        
        #if left and right branches values is same, assign ans to dictionary
        if yes_answer == no_answer:
            sub_tree = yes_answer
        #else return the sub_tree dictionary.
        else:
            sub_tree[question].append(yes_answer)
            sub_tree[question].append(no_answer)
            
    #return final decision tree as a dictionary
    return sub_tree

## Test out the decision tree classifier model splitting by info gain:

In [33]:
#split by info gain

#Using train set as input, create decision tree model splitting by info gain
#Set max_depth = 10, min_samples = 10
#the maximum depth of the decision tree/the maximum number of splits
#min_samples: the minimum number of samples required for a split to occur
#max_depth and min_samples may be tuned if needed for higher accuracy.
tree_info_gain = decision_tree_classifier_infogain(train_df, max_depth=10, min_samples=10)

#Predict class labels for test set
#Get a list of predictions for all rows in test_df
prediction_info_gain = predict(test_df, tree_info_gain)

In [34]:
#View tree
pprint(tree_info_gain)

{'marital-status = Married-civ-spouse': [{'capital-gain = 0': [{'occupation = Exec-managerial': [{'education = Masters': [{'relationship = Asian-Pac-Islander': ['<=50K',
                                                                                                                                                                 {'age = 28': ['<=50K',
                                                                                                                                                                               {'age = 34': ['<=50K',
                                                                                                                                                                                             {'hours-per-week = 45': [{'age = 29': ['<=50K',
                                                                                                                                                                                                                   

In [35]:
#Calc accuracy of dt classifier by comparing prediction to actual class label
#number of correct predictions made/total number of predictions made.
calculate_accuracy(test_df, tree_info_gain)

0.8513883133029424

#### An accuracy of 0.8513883133029424 for a decision tree classifier means that the model correctly predicted the class label for 85.13% of the samples in the test data when splitting by information gain. Hence, the tree is very accurate in predicting income.

## 2. decision tree classifier based on gini index

In [36]:
#parameters
#df: the dataframe containing the data to be trained and classified
#max_depth: the maximum depth of the decision tree/the maximum number of splits
#min_samples: the minimum number of samples required for a split to occur
#counter: a counter variable to keep track of the depth of the tree
def decision_tree_classifier_gini_index(df, max_depth, min_samples, counter=0):
    
    #initialize variables at the start
    if counter == 0:
        global COLUMN_HEADERS
        COLUMN_HEADERS = df.columns
        data = df.values
    else:
        data = df           
    
    # early stopping conditions: 
    # if the data is a leaf node/min samples > number of samples in data/max depth reached = get leaf node classification
    if (check_leaf(data)) or (len(data) < min_samples) or (counter == max_depth):
        classification = classify_data(data)
        
        return classification
    
    #Else, get all possible splits and choose the best split with best info gain
    #Then it splits the data into two branches, left and right, based on the best split and the question generated from it.
    else:    
        #increment the counter for each iteration
        counter += 1
        #get all possible splits
        possible_splits = get_possible_splits(data)
        
        #returns the feature/col with highest gini index
        split_column, split_value = determine_best_split_gini_index(data, possible_splits)
        
        #split data into left and right branches using best feature as decided by gini index
        left, right = split_data(data, split_column, split_value)
        
        # determine question (key)
        # get the name of the feature that was used for the best split
        feature_name = COLUMN_HEADERS[split_column]
            
        #format question (key) (e.g. "feature_name = split_value")
        question = "{} = {}".format(feature_name, split_value)
        
        #initialise dictionary for storing tree
        sub_tree = {question: []}
    
        #recursive call with left and right branches
        yes_answer = decision_tree_classifier_gini_index(left, max_depth, min_samples, counter)
        no_answer = decision_tree_classifier_gini_index(right, max_depth, min_samples, counter)
        
        #if left and right branches values is same, assign ans to dictionary
        if yes_answer == no_answer:
            sub_tree = yes_answer
        #else return the sub_tree dictionary.
        else:
            sub_tree[question].append(yes_answer)
            sub_tree[question].append(no_answer)
        
    #return final decision tree as a dictionary
    return sub_tree

## Test out the decision tree classifier model splitting by gini index:

In [37]:
#split by gini index

#Using training set as input, create decision tree model splitting by gini index
#Set max_depth = 10, min_samples = 10
#the maximum depth of the decision tree/the maximum number of splits
#min_samples: the minimum number of samples required for a split to occur
#max_depth and min_samples may be tuned if needed for higher accuracy.
tree_gini_index = decision_tree_classifier_gini_index(train_df, max_depth = 10, min_samples = 10)

#Predict class labels for test set
#Get a list of predictions for all rows in test_df
prediction_gini_index = predict(test_df, tree_gini_index)

In [38]:
#View tree
pprint(tree_gini_index)

{'marital-status = Married-civ-spouse': [{'occupation = Exec-managerial': [{'capital-gain = 0': [{'workclass = Self-emp-not-inc': [{'capital-loss = 0': [{'education = Bachelors': [{'hours-per-week = 50': ['<=50K',
                                                                                                                                                                                                             {'age = 25': ['<=50K',
                                                                                                                                                                                                                           {'hours-per-week = 40': [{'age = 44': ['<=50K',
                                                                                                                                                                                                                                                                  '>50K']},
                       

In [39]:
#Calc accuracy of dt classifier by comparing prediction to actual class label
#number of correct predictions made/total number of predictions made.
calculate_accuracy(test_df, tree_gini_index)

0.8509738914214671

#### An accuracy of 0.8509738914214671 for a decision tree classifier means that the model correctly predicted the class label for 85.09% of the samples in the test data when splitting by gini index. Hence, the tree is very accurate in predicting income.  However, it is slightly less accurate than predicting with info gain.

## 3. Decision Tree Classifier based on gain ratio

In [40]:
#parameters
#df: the dataframe containing the data to be trained and classified
#max_depth: the maximum depth of the decision tree/the maximum number of splits
#min_samples: the minimum number of samples required for a split to occur
#counter: a counter variable to keep track of the depth of the tree
def decision_tree_classifier_gain_ratio(df, max_depth, min_samples, counter=0):
    
    #initialize variables at the start
    if counter == 0:
        global COLUMN_HEADERS
        COLUMN_HEADERS = df.columns
        data = df.values
    else:
        data = df     
    
    # early stopping conditions: 
    # 1. if data is a leaf node
    # 2. if number of samples in data is less than min_samples
    # 3. if the maximum depth has been reached
    if (check_leaf(data)) or (len(data) < min_samples) or (counter == max_depth):
        
        # classify the data as the leaf node classification
        classification = classify_data(data)
        
        return classification
    
    #Else, get all possible splits and choose the best split with best gain ratio
    #Then it splits the data into two branches, left and right, based on the best split and the question generated from it.
    else:    
        # increment the counter for each iteration
        counter += 1
        # get all possible splits
        possible_splits = get_possible_splits(data)
        
        # determine the best split based on the best gain ratio
        split_column, split_value = determine_best_split_gain_ratio(data, possible_splits)
        
        #split data into left and right branches using best feature as decided by gain ratio
        left, right = split_data(data, split_column, split_value)
        
        # determine question (key)
        # get the name of the feature that was used for the best split
        feature_name = COLUMN_HEADERS[split_column]
            
        #format question (key) (e.g. "feature_name = split_value")
        question = "{} = {}".format(feature_name, split_value)
        
        #initialise dictionary for storing tree
        sub_tree = {question: []}
        
        #recursive call with left and right branches
        yes_answer = decision_tree_classifier_gain_ratio(left, max_depth, min_samples, counter)
        no_answer = decision_tree_classifier_gain_ratio(right, max_depth, min_samples, counter)
    
        #if left and right branches values is same, assign ans to dictionary
        if yes_answer == no_answer:
            sub_tree = yes_answer
        
        #else append the sub_tree dictionary.
        else:
            sub_tree[question].append(yes_answer)
            sub_tree[question].append(no_answer)
        
    #return final decision tree as a dictionary
    return sub_tree

## Test out the decision tree classifier model splitting by gain ratio:

In [41]:
#split by gain ratio

#Using training set as input, create decision tree model splitting by gain ratio
#Set max_depth = 10, min_samples = 10
#the maximum depth of the decision tree/the maximum number of splits
#min_samples: the minimum number of samples required for a split to occur
#max_depth and min_samples may be tuned if needed for higher accuracy.
tree_gain_ratio = decision_tree_classifier_gain_ratio(train_df, max_depth = 10, min_samples = 10)

#Predict class labels for test set
#Get a list of predictions for all rows in test_df
prediction_gain_ratio = predict(test_df, tree_gain_ratio)

In [42]:
#View tree
pprint(tree_gain_ratio)

{'marital-status = Married-civ-spouse': [{'occupation = Exec-managerial': [{'workclass = Self-emp-not-inc': [{'education = Bachelors': [{'age = 54': ['>50K',
                                                                                                                                                      {'age = 25': ['<=50K',
                                                                                                                                                                    {'age = 26': ['<=50K',
                                                                                                                                                                                  {'age = 39': ['<=50K',
                                                                                                                                                                                                {'age = 40': ['<=50K',
                                                                  

In [43]:
#Calc accuracy of dt classifier by comparing prediction to actual class label
#number of correct predictions made/total number of predictions made.
calculate_accuracy(test_df, tree_gain_ratio)

0.8406962287608786

#### An accuracy of 0.8406962287608786 for a decision tree classifier means that the model correctly predicted the class label for 84.06% of the samples in the test data when splitting by gain ratio. Hence, the tree is very accurate in predicting income. However, it is  slightly less accurate than predicting with gini index/info gain.

## 4. Ensemble method to build new classifier

In [44]:
#Utilises voting function on prediction outcome of the 3 dt models before.

#Parameters
#prediction: an array containing the predictions of an individual model (info gain)
#prediction1: an array containing the predictions of an individual model (gini index)
#prediction2: an array containing the predictions of an individual model (gain ratio)
def ensemble_voting(prediction, prediction1, prediction2):

    #Add all models and predictions to dataframe
    all_predictions = pd.DataFrame({'model1': prediction,
                           'model2': prediction1,
                           'model3': prediction2})
    
    #Find highest occurance per column (Voting function)
    ensemble_predictions = all_predictions.mode(axis=1)
    
    #returns a numpy array containing the ensemble predictions
    return np.array(ensemble_predictions[0])

In [45]:
# Get the ensemble predictions
prediction_ensemble = ensemble_voting(prediction_info_gain, prediction_gini_index, prediction_gain_ratio)

## Test out the accuracy of the ensemble method:

In [46]:
#Test accuracy of ensemble function for analysis
def test_ensemble_accuracy (df, ensemble_pred):
    
    #If classification for ensemble_pred is correct, add it to df["classification_correct"]
    df["classification_correct"] = ensemble_pred == df["income"]
    
    # Calculate the accuracy using mean of total predictions
    accuracy = df["classification_correct"].mean()
    
    return accuracy
            
test_ensemble_accuracy (test_df, prediction_ensemble)

0.8498135101533361

#### An accuracy of 0.8498135101533361 for the ensemble function means that the function correctly predicted the class label for 84.98% of the samples in the test data when utilising a voting function on the prediction outcomes of the three Decision Tree (DT) models. It indicates that the majority of the predictions made by the 3 decision trees are correct. 

#### The accuracy of the ensemble function is slightly lower than the DT models splitting by information gain (85.13%), very similar to the gini index model (85.09%), and is higher in accuracy compared to the DT model splitting by gain ratio (84.06%). As the predictions are aggregated to form a prediction. 

#### The accuracy of the model is very close to the information gain and gini index model due to the aggregation of predictions and voting function. However, the lowered accuracy of the gain ratio model might have affected the performance of the ensemble method. The combined predictions might not always be better than the best individual model, and the accuracy could suffer if the models have conflicting or complementary weaknesses. As such, a voting function might not always be the most accurate. Additionally, if the models in the ensemble are not diverse enough, the accuracy might not improve significantly with the ensemble method.